File: source/kotlin/kotlin-native/Parallel.kt
// Parallel map with atomic work stealing (Kotlin/Native)

import kotlin.concurrent.AtomicInt
import kotlin.native.concurrent.TransferMode
import kotlin.native.concurrent.Worker
import kotlin.random.Random

object Parallel {
    private class Context<T, R>(
        val tasks: Array<T>,
        val order: IntArray,
        val next: AtomicInt,
        val results: Array<Any?>,
        val run: (T) -> R,
    )

    private fun <T, R> runWorker(ctx: Context<T, R>) {
        val count = ctx.tasks.size
        var index = ctx.next.incrementAndGet() - 1
        while (index < count) {
            val idx = ctx.order[index]
            ctx.results[idx] = ctx.run(ctx.tasks[idx])
            index = ctx.next.incrementAndGet() - 1
        }
    }

    fun <T, R> map(cores: Int, tasks: Array<T>, run: (T) -> R): List<R> {
        val count = tasks.size
        val results = arrayOfNulls<Any>(count)

        // Shuffle work order for load balancing
        val order = IntArray(count) { it }
        val rng = Random.Default
        for (i in count - 1 downTo 1) {
            val j = rng.nextInt(i + 1)
            val tmp = order[i]; order[i] = order[j]; order[j] = tmp
        }

        val ctx = Context(tasks, order, AtomicInt(0), results, run)

        val workers = List(cores - 1) { Worker.start() }
        val futures = workers.map { w ->
            w.execute(TransferMode.UNSAFE, { ctx }) { runWorker(it) }
        }

        runWorker(ctx)
        futures.forEach { it.result }
        workers.forEach { it.requestTermination().result }

        @Suppress("UNCHECKED_CAST")
        return results.map { it as R }
    }
}

File: source/F#/src/parallel.fs
module Parallel

type Env = System.Environment
type Atomic = System.Threading.Interlocked
type Task = System.Threading.Tasks.Task
type Thread = System.Threading.Thread

let swap (arr : _[]) i j =
   let temp = arr.[i]
   arr.[i] <- arr.[j]
   arr.[j] <- temp

let map
   (cores : int)
   (init : 'b)
   (run : 'a -> 'b)
   (tasks : 'a list)
   : 'b list =
   let tasks = Array.ofList tasks
   let count = tasks.Length

   let mutable next = -1
   let order = Array.init count id

   let results = Array.create count init

   let shuffle arr =
      for i = count - 1 downto 1 do
         let j = System.Random.Shared.Next (i + 1)
         swap arr i j

   shuffle order

   let worker () =
      let rec loop () =
         let index = Atomic.Increment (&next)

         if index < count then
            let index = order.[index]
            results.[index] <- run tasks.[index]
            loop ()

      loop ()

   let workers = [| for _ in 1 .. cores - 1 -> Task.Run (worker) |]

   worker ()
   Task.WaitAll workers
   Array.toList results
